tally/include/tally/buckets_iterator.h

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#pragma once
+
+#include <iterator>
+
+#include "tally/src/buckets_calculator.h"
+
+namespace tally {
+
+class BucketsIterator : public std::iterator<std::input_iterator_tag, double> {
```

Comment:
OK just to be clear about iterator types:

+ InputIterator: "An InputIterator is an Iterator that can read from the pointed-to element. InputIterators only guarantee validity for single pass algorithms: once an InputIterator i has been incremented, all copies of its previous value may be invalidated." (https://en.cppreference.com/w/cpp/named_req/InputIterator)
+ RandomAccessIterator: "A ForwardIterator is an Iterator that can read data from the pointed-to element. Unlike InputIterator and OutputIterator, it can be used in multipass algorithms." (https://en.cppreference.com/w/cpp/named_req/ForwardIterator)
+ BidirectionalIterator: "A BidirectionalIterator is a ForwardIterator that can be moved in both directions (i.e. incremented and decremented)." (https://en.cppreference.com/w/cpp/named_req/BidirectionalIterator)
+ RandomAccessIterator: "A RandomAccessIterator is a BidirectionalIterator that can be moved to point to any element in constant time. A pointer to an element of an array satisfies all requirements of RandomAccessIterator" (https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator)

If you're not using a file/socket, it isn't an input iterator. Seeing as you are using a container, I'd expect this to be a random access iterator. What I usually do to avoid making my own iterator type is to use something like `using Iterator = underlying_container::iterator` in my own class and expose the iterators of the field directly. Not sure exactly how you are getting iterator value here though so it may not be relevant.

Reply:
The reason I chose to use an InputIterator is that it has the smallest API. The iterator is really just a wrapper around an index so it should be straightforward to implement the other iterators but it should only be used by clients when making a single pass over all buckets so I didn't want to support more methods than necessary.

Reply:
There's no underlying iterator for the calculator, it just returns outputs a number based on the current index.

Reply:
It's okay for an internal API, but the stl algorithms require different iterators for different functions. For example, I highly doubt you can sort these iterators.